public/app/features/explore/state/query.ts

```diff
@@ -702,6 +701,103 @@ export const runQueries = createAsyncThunk<void, RunQueriesOptions>(
+    const { defaultCorrelationEditorDatasource, scopedVars, showCorrelationEditorLinks } = await getCorrelationsData(
+      getState(),
+      exploreId
+    );
+    const correlations$ = getCorrelations(exploreId);
+
+    let newQuerySource: Observable<ExplorePanelData>;
+
+    const logQueries = queryResponse.logsResult?.queries || [];
+    const queries = logQueries.map((query: DataQuery) => ({
+      ...query,
+      datasource: query.datasource || datasourceInstance?.getRef(),
+    }));
+
+    if (!hasNonEmptyQuery(queries) || !datasourceInstance) {
+      return;
+    }
+
+    // Some datasource's query builders allow per-query interval limits,
+    // but we're using the datasource interval limit for now
+    const minInterval = datasourceInstance?.interval;
+
+    const queryOptions: QueryOptions = {
+      minInterval,
+      // maxDataPoints is used in:
+      // Loki - used for logs streaming for buffer size, with undefined it falls back to datasource config if it supports that.
+      // Elastic - limits the number of datapoints for the counts query and for logs it has hardcoded limit.
+      // Influx - used to correctly display logs in graph
+      // TODO:unification
+      // maxDataPoints: mode === ExploreMode.Logs && datasourceId === 'loki' ? undefined : containerWidth,
+      maxDataPoints: containerWidth,
+    };
+
+    const timeZone = getTimeZone(getState().user);
+    const range = getTimeRange(
+      timeZone,
+      {
+        from: dateTimeForTimeZone(timeZone, absoluteRange.from),
+        to: dateTimeForTimeZone(timeZone, absoluteRange.to),
+      },
+      getFiscalYearStartMonth(getState().user)
+    );
+    const transaction = buildQueryTransaction(exploreId, queries, queryOptions, range, false, timeZone, scopedVars);
+
+    dispatch(changeLoadingStateAction({ exploreId, loadingState: LoadingState.Loading }));
+
+    newQuerySource = combineLatest([runRequest(datasourceInstance, transaction.request), correlations$]).pipe(
+      mergeMap(([data, correlations]) =>
+        decorateData(
+          // Query splitting, otherwise duplicates results
+          data.state === LoadingState.Done ? mergeDataSeries(queryResponse, data) : data,
+          queryResponse,
+          absoluteRange,
+          undefined,
+          queries,
+          correlations,
+          showCorrelationEditorLinks,
+          defaultCorrelationEditorDatasource
+        )
+      )
+    );
+
+    newQuerySource.subscribe({
+      next(data) {
+        if (data.logsResult !== null && data.state === LoadingState.Done) {
+          reportInteraction('grafana_explore_logs_scrolled', {
+            datasourceType: datasourceInstance.type,
+          });
+        }
+        dispatch(queryStreamUpdatedAction({ exploreId, response: data }));
+      },
+      error(error) {
+        dispatch(notifyApp(createErrorNotification('Query processing error', error)));
```

Comment:
There's still a lot of code duplication that would great to group together. There are no tests added to cover this and  no guard to ensure changes in one place are applied in both places and can lead to bugs.

Reply:
Compared both actions side by side, and roughly what is not commented is what they have in common, which would be very awkward or challenging to abstract:

![imagen](https://github.com/grafana/grafana/assets/1069378/f95f607f-0332-4384-a464-c2ce07505bd4)

I will add an item for the next iteration before GA to make sure we have enough coverage, specially of the new action. Honestly, the fact that it's Redux makes it a bit harder to properly test.